runtime/io/common/src/aws/smithy/kotlin/runtime/io/CompletingSource.kt

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+import aws.smithy.kotlin.runtime.util.encodeBase64String
+import kotlinx.coroutines.CompletableDeferred
+
+/**
+ * An [SdkSource] which uses the digest of an underlying [HashingSource] to complete a [CompletableDeferred]
+ * @param deferredChecksum The [CompletableDeferred] to be completed when the underlying source is exhausted
+ * @param source the [HashingSource] which will be digested and used to complete [deferredChecksum]
+ */
+@InternalApi
+public class CompletingSource(private val deferredChecksum: CompletableDeferred<String>, private val source: HashingSource) : SdkSource by source {
```

Comment:
**comment**: Same comment, this is so simple of wrapping code I'm not sure it needs to be made generic. It could just live with the as an internal implementation detail.